File: dotnet/Lexy/Lexy.Compiler/Language/Expressions/CaseExpression.cs
using System.Collections.Generic;
using Lexy.Compiler.Parser;

namespace Lexy.Compiler.Language.Expressions
{
    public class CaseExpression : Expression, IParsableNode, IDependantExpression
    {
        private readonly ExpressionList expressions;

        public Expression Value { get; }
        public IEnumerable<Expression> Expressions => expressions;
        public bool IsDefault { get; }

        private CaseExpression(Expression value, bool isDefault, ExpressionSource source, SourceReference reference) : base(source, reference)
        {
            Value = value;
            IsDefault = isDefault;
            expressions = new ExpressionList(reference);
        }

        public static ParseExpressionResult Parse(ExpressionSource source)
        {
            var tokens = source.Tokens;
            if (!IsValid(tokens))
            {
                return ParseExpressionResult.Invalid<IfExpression>("Not valid.");
            }

            if (tokens.IsKeyword(0, Keywords.Default))
            {
                return ParseDefaultCase(source, tokens);
            }

            if (tokens.Length == 1)
            {
                return ParseExpressionResult.Invalid<CaseExpression>("Invalid 'case'. No parameters found.");
            }

            var value = tokens.TokensFrom(1);
            var valueExpression = ExpressionFactory.Parse(source.File, value, source.Line);
            if (valueExpression.Status == ParseExpressionStatus.Failed) return valueExpression;

            var reference = source.CreateReference();

            var expression = new CaseExpression(valueExpression.Expression, false , source, reference);

            return ParseExpressionResult.Success(expression);
        }

        private static ParseExpressionResult ParseDefaultCase(ExpressionSource source, TokenList tokens)
        {
            if (tokens.Length != 1)
            {
                return ParseExpressionResult.Invalid<CaseExpression>("Invalid 'default' case. No parameters expected.");
            }

            var reference = source.CreateReference();
            var expression = new CaseExpression(null, true, source, reference);
            return ParseExpressionResult.Success(expression);
        }

        public static bool IsValid(TokenList tokens)
        {
            return tokens.IsKeyword(0, Keywords.Case)
                   || tokens.IsKeyword(0, Keywords.Default);
        }

        public IParsableNode Parse(IParserContext context)
        {
            var line = context.CurrentLine;
            if (line.IsComment() || line.IsEmpty())
            {
                return this;
            }

            var expression = ExpressionFactory.Parse(context.SourceCode.File, line.Tokens, line);
            if (expression.Status == ParseExpressionStatus.Failed)
            {
                context.Logger.Fail(context.LineStartReference(), expression.ErrorMessage);
                return null;
            }

            expressions.Add(expression.Expression, context);
            return expression.Expression is IParsableNode node ? node : this;
        }

        public override IEnumerable<INode> GetChildren()
        {
            if (Value != null) yield return Value;

            yield return expressions;
        }

        protected override void Validate(IValidationContext context)
        {
        }

        public override VariableType DeriveType(IValidationContext context) => Value?.DeriveType(context);

        public void LinkPreviousExpression(Expression expression, IParserContext context)
        {
            if (!(expression is SwitchExpression switchExpression))
            {
                context.Logger.Fail(Reference, "'case' should be following a 'switch' statement. No 'switch' statement found.");
                return;
            }

            switchExpression.LinkElse(this);
        }
    }
}

File: typescipt/src/language/expressions/caseExpression.ts
import {Expression} from "./Expression";
import {VariableType} from "../types/variableType";

export class CaseExpression extends Expression, IParsableNode, IDependantExpression {



  private readonly ExpressionList expressions;

  public nodeType: "CaseExpression"
   public Expression Value
   public Array<Expression> Expressions => expressions;
   public boolean IsDefault

   private CaseExpression(Expression value, boolean isDefault, ExpressionSource source, SourceReference reference) : base(
     source, reference) {
     Value = value;
     IsDefault = isDefault;
     expressions = new ExpressionList(reference);
   }

   public linkPreviousExpression(expression: Expression, context: IParseLineContext): void {
     if (expression is not SwitchExpression switchExpression) {
       context.logger.fail(this.reference,
         `'case' should be following a 'switch' statement. No 'switch' statement found.`);
       return;
     }

     switchExpression.LinkElse(this);
   }

   public parse(context: IParseLineContext): IParsableNode {
     let expression = expressions.parse(context);
     return expression.result is IParsableNode node ? node : this;
   }

   public override getChildren(): Array<INode> {
     if (Value != null) yield return Value;

     yield return expressions;
   }

   public static parse(source: ExpressionSource): ParseExpressionResult {
     let tokens = source.tokens;
     if (!IsValid(tokens)) return newParseExpressionFailed(IfExpression>(`Not valid.`);

     if (tokens.isKeyword(0, Keywords.Default)) return ParseDefaultCase(source, tokens);

     if (tokens.length == 1)
       return newParseExpressionFailed(CaseExpression>(`Invalid 'case'. No parameters found.`);

     let value = tokens.tokensFrom(1);
     let valueExpression = ExpressionFactory.parse(value, source.line);
     if (!valueExpression.state != 'success') return valueExpression;

     let reference = source.createReference();

     let expression = new CaseExpression(valueExpression.result, false, source, reference);

     return newParseExpressionSuccess(expression);
   }

   private static parseDefaultCase(source: ExpressionSource, tokens: TokenList): ParseExpressionResult {
     if (tokens.length != 1)
       return newParseExpressionFailed(CaseExpression>(`Invalid 'default' case. No parameters expected.`);

     let reference = source.createReference();
     let expression = new CaseExpression(null, true, source, reference);
     return newParseExpressionSuccess(expression);
   }

   public static isValid(tokens: TokenList): boolean {
     return tokens.isKeyword(0, Keywords.Case)
        || tokens.isKeyword(0, Keywords.Default);
   }

   protected override validate(context: IValidationContext): void {
   }

   public override deriveType(context: IValidationContext): VariableType | null {
     return Value?.deriveType(context);
   }
}
